### PR TITLE
Working on voucherify poc

### PR DESCRIPTION
**Pull request by marcin-slezak**

### CODE REVIEW

File: src/voucherify-integration/contentfulApi.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+      ? `https://cdn.contentful.com/spaces/${process.env.REACT_APP_CONTENTFUL_SPACE_ID}/environments/master/entries/${entryId}?access_token=${process.env.REACT_APP_CONTENTFUL_DELIVERY_API_ACCESS_TOKEN}`
+      : `https://preview.contentful.com/spaces/${process.env.REACT_APP_CONTENTFUL_SPACE_ID}/environments/master/entries/${entryId}?access_token=${process.env.REACT_APP_CONTENTFUL_PREVIEW_API_ACCESS_TOKEN}`
+
+  const response = await fetch(url, {
+    method: 'GET',
+  })
+  try {
+    return (await response.json())?.fields || {}
+  } catch (e) {
+    console.log(e)
```
Comment (marcin-slezak) [inline src/voucherify-integration/contentfulApi.js:13]: So we allow the application to continue?

File: src/voucherify-integration/emporixApi.js [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+    `https://api.emporix.io/cart/piotr/carts/${cartId}/discounts`,
+    {
+      method: 'Delete',
+      headers: {
+        Authorization: `Bearer ${await getEmporixAPIAccessToken()}`,
+      },
+    }
+  )
+  if (resultRaw.status !== 204) {
+    console.log({ error: `coupons could not be removed from cart` })
```
Comment (marcin-slezak) [inline src/voucherify-integration/emporixApi.js:23]: ignored errors?

File: src/voucherify-integration/emporixApi.js
Comment (marcin-slezak) [inline src/voucherify-integration/emporixApi.js:1]: error handling sucks here

File: src/voucherify-integration/emporixApi.js [hunk 3]
```diff
@@ -0,0 +1,154 @@
+import { getEmporixAPIAccessToken } from './getEmporixAPIAccessToken'
```
Comment (marcin-slezak) [inline src/voucherify-integration/emporixApi.js:1]: Definitely, you can rethink how looks the architecture of your code. You have a service that uses `getEmporixAPIAccessToken` which inside uses local storage for cache 🥶

File: src/voucherify-integration/voucherifyApi.js [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+        limit: 100,
+        ignore_rules_cache: true, // hidden,
+        starting_after: cursor,
+        filters: {},
+      },
+    },
+    path: 'qualifications',
+    method: 'POST',
+  })
+  if (qualificationsResponse.status !== 200) {
```
Comment (marcin-slezak) [inline src/voucherify-integration/voucherifyApi.js:47]: Ignored error?

File: src/context/cart-provider.js [hunk 5]
```diff
@@ -104,6 +106,10 @@ const CartProvider = ({ children }) => {
     return cartAccount.discounts
   }, [cartAccount.discounts])
 
+  const mixins = useMemo(() => {
```
[the remark this answers is not in the document]
Comment (p-zielinski) [reply src/context/cart-provider.js:109]: emporix free to use metadata naming

File: src/pages/cart/CartPage.js [hunk 6]
```diff
@@ -1,27 +1,391 @@
...
+    )
+    setAllOtherQualifications(
+      qualificationsAllScenario.filter(
+        (qualification) => !qualificationsIdsSoFar.includes(qualification.id)
+      )
+    )
+  }
+
+  const setBundleQualificationsEnriched = async (bundleQualifications) => {
+    setBundleQualifications(bundleQualifications)
```
[the remark this answers is not in the document]
Comment (p-zielinski) [reply src/pages/cart/CartPage.js:70]: yes, this is as I intended.
1. I want to show promotion ASAP
2. Once I find what products shall be added to use the promotion we shall update it

[the remark this answers is not in the document]
Comment (p-zielinski) [reply src/pages/cart/CartPage.js:70]: I'll add com

File: src/pages/cart/CartPage.js [hunk 7]
```diff
@@ -1,27 +1,391 @@
...
+    const enrichedBundleQualifications = await asyncMap(
+      bundleQualifications,
+      async (bundleQualification) => {
+        const allValidationRuleIds = (
+          bundleQualification?.validation_rule_assignments?.data || []
+        )
+          .map((validationRule) => validationRule.rule_id)
+          .filter((e) => e)
+        const allValidationRules = await Promise.all(
+          await asyncMap(allValidationRuleIds, (id) => getValidationRule(id))
```
[the remark this answers is not in the document]
Comment (p-zielinski) [reply src/pages/cart/CartPage.js:80]: 1 promotion tier with 1 validation rule
that make 2 requests.

File: src/pages/checkout/CheckoutPage.js [hunk 8]
```diff
@@ -56,224 +64,188 @@ const ReviewOrderAction = ({ action }) => {
...
+      variant="outlined"
+      onClick={applyPromotionOnCart}
+    />
+  )
+}
+
+export const Coupon = () => {
+  const [code, setCode] = useState('')
+  const [error, setError] = useState('')
+  const [keyValue, setKeyValue] = useState(Math.random())
```
[the remark this answers is not in the document]
Comment (p-zielinski) [reply src/pages/checkout/CheckoutPage.js:104]: It's a fix. Unfortunately I could not find anything better than hard component refresh.

File: src/voucherify-integration/cartUpdateActions.js [hunk 9]
```diff
@@ -0,0 +1,53 @@
+export class CartUpdateActions {
```
[the remark this answers is not in the document]
Comment (p-zielinski) [reply src/voucherify-integration/cartUpdateActions.js:1]: colect data :)

File: src/voucherify-integration/contentfulApi.js (hunk 1 above)
Comment (p-zielinski) [reply src/voucherify-integration/contentfulApi.js:13]: yes, that is not a fatal error. Just some entry was deleted or moved from published to unpublished or something.

File: src/voucherify-integration/emporixApi.js (hunk 2 above)
Comment (p-zielinski) [reply src/voucherify-integration/emporixApi.js:40]: shall we try again?

File: src/voucherify-integration/emporixApi.js (hunk 3 above)
Comment (p-zielinski) [reply src/voucherify-integration/emporixApi.js:1]: that is why noone shall do server on a frontend

File: src/voucherify-integration/getEmporixAPIAccessToken.js [hunk 10]
```diff
@@ -0,0 +1,41 @@
+export const getEmporixAPIAccessToken = async () => {
+  const accessToken = localStorage.getItem('emporixAccessToken')
```
[the remark this answers is not in the document]
Comment (p-zielinski) [reply src/voucherify-integration/getEmporixAPIAccessToken.js:2]: Disagree.

File: src/voucherify-integration/getDiscountsValues.js [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+  })
+
+  return discounts.sort((discount1, discount2) => {
+    if (discount1?.target?.predicate === 'true') return 1
+    if (discount2?.target?.predicate === 'true') return -1
+    return -1
+  })
+}
+
+const applyDiscount = (discounts, items) => {
```
[the remark this answers is not in the document]
Comment (p-zielinski) [reply src/voucherify-integration/getDiscountsValues.js:44]: where? We do no mutate input param here.

File: README-VOUCHERIFY.md [hunk 12]
```diff
@@ -0,0 +1,166 @@
+# Emporix B2B Showcase
+
```
Comment (marcin-slezak) [inline README-VOUCHERIFY.md:2]: Table of contents + numeration + reasonable structure, e.g.
- introduction
  - Contentful role in this integration 
  - Qualifications
- how to prepare the environment
  - prepare Emporix, Voucherify and Contentful services 
    - migrations
    - how to get API keys
  - how to prepare showcase application
    - requirements (node, npm, ?)
    - installation process
       - download repo
       - install dependencies
       - configure .env
       -  run
    - commands? (yarn start / yarn install)
- how it works
  - Applying vouchers and promotion tiers 
  - Bundles
  - Integration from a technical perspective

File: README-VOUCHERIFY.md [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+about that voucher, so you will try to apply later on (on cart page).
+
+### Bundles
+
+We implemented additional functionality to bundle type of promotions. You may add missing products to your cart directly from promotion holder.
+A missing product is a product that is missing to receive a discount. *(readme-images/Bundles-1.png)*
+
+## Pre Requirements 
+
+### General
```
Comment (marcin-slezak) [inline README-VOUCHERIFY.md:55]: General?

File: src/integration/emporix/getEmporixAPIAccessToken.js [hunk 14]
```diff
@@ -0,0 +1,43 @@
+export const getEmporixAPIAccessToken = async () => {
+  //This shall be stored in memory cache.
+  //However, this is PoC based on React, so we must use localstorage to not lose data, after we close a cart.
```
Comment (marcin-slezak) [inline src/integration/emporix/getEmporixAPIAccessToken.js:3]: check if it works for you


```ks
let accessToken;
let emporixAccessTokenExpiresAt;

export const getEmporixAPIAccessToken = async () => {
  if (
    !isNaN(expiresAt) &&
    typeof expiresAt === 'number' &&
    new Date().getTime() <= expiresAt
  ) {
    return accessToken
  }
  // ...
    emporixAccessToken = newAccessToken;
    emporixAccessTokenExpiresAt = newExpiresAt
}
```

File: README-VOUCHERIFY.md [hunk 15]
```diff
@@ -0,0 +1,166 @@
...
+      - EachProductRow.js
+      - ProductDetailPage.js
+      - ProductListContent.js
+      - ProductPage.js
+      - index.js
+  - services
+    - cart.service.js
+- yarn.lock
+
+Additionally, we added some files, please see PR https://github.com/rspective/emporix-b2b-showcase/pull/5/files
```
[the remark this answers is not in the document]
Comment (p-zielinski) [reply README-VOUCHERIFY.md]: yes, shall we do someting sbout it?

File: src/integration/voucherify/getContentfulEntryFields.js [hunk 16]
```diff
@@ -0,0 +1,19 @@
...
+      ? `https://cdn.contentful.com/spaces/${process.env.REACT_APP_CONTENTFUL_SPACE_ID}/environments/master/entries/${entryId}?access_token=${process.env.REACT_APP_CONTENTFUL_DELIVERY_API_ACCESS_TOKEN}`
+      : `https://preview.contentful.com/spaces/${process.env.REACT_APP_CONTENTFUL_SPACE_ID}/environments/master/entries/${entryId}?access_token=${process.env.REACT_APP_CONTENTFUL_PREVIEW_API_ACCESS_TOKEN}`
+
+  const response = await fetch(url, {
+    method: 'GET',
+  })
+  try {
+    return (await response.json())?.fields || {}
+  } catch (e) {
+    //someone had to remove the entry
```
[the remark this answers is not in the document]
Comment (p-zielinski) [reply src/integration/voucherify/getContentfulEntryFields.js:13]: added proper error handling

File: src/integration/voucherify/cartUpdateActions.js [hunk 17]
```diff
@@ -0,0 +1,53 @@
+export class CartUpdateActions {
```
[the remark this answers is not in the document]
Comment (p-zielinski) [reply src/integration/voucherify/cartUpdateActions.js]: deleted

**Comment by p-zielinski** [comment]
Voucherify PoC is ready

**Pull request merged**